xlive/H2MOD/Variants/H2X/H2X.cpp

```diff
@@ -43,91 +42,47 @@ void H2X::ApplyMapLoadPatches(bool enable)
 {
 	for (auto& weapon : weapons)
 	{
-		auto required_datum = tags::find_tag(blam_tag::tag_group_type::weapon, weapon.tag_string);
-		BYTE* weapon_tag = tags::get_tag<blam_tag::tag_group_type::weapon, BYTE>(required_datum);
-		if (weapon_tag != nullptr)
-			{
-				int barrel_data_block_size = 236;
-				tags::tag_data_block* barrel_data_block = reinterpret_cast<tags::tag_data_block*>(weapon_tag + 720);
-
-				if (barrel_data_block->block_data_offset != -1)
-				{
-					*(float*)(tags::get_tag_data()
-						+ barrel_data_block->block_data_offset
-						+ barrel_data_block_size * weapon.barrel_data_block_index
-						+ (weapon.rounds_per_second_based ? 8 : 32)) = (enable ? weapon.h2x_rate_of_fire : weapon.original_rate_of_fire);
-			}
+		datum weapon_datum = tags::find_tag(blam_tag::tag_group_type::weapon, weapon.tag_string);
+		if (weapon_datum != DATUM_INDEX_NONE)
+		{
+			s_weapon_group_definition* weapon_tag = tags::get_tag_fast<s_weapon_group_definition>(weapon_datum);
+			weapon.rounds_per_second_based ?
+				weapon_tag->barrels[weapon.barrel_data_block_index]->rounds_per_second_upper = (enable ? weapon.h2x_rate_of_fire : weapon.original_rate_of_fire)
+				: weapon_tag->barrels[weapon.barrel_data_block_index]->fire_recovery_time = (enable ? weapon.h2x_rate_of_fire : weapon.original_rate_of_fire);
 		}
 	}
 
-	if (enable && !Memory::IsDedicatedServer() && h2mod->GetEngineType() == e_engine_type::_multiplayer)
+	if (enable && !Memory::IsDedicatedServer() && s_game_globals::game_is_multiplayer())
```

Comment:
`h2mod->GetEngineType() == e_engine_type::_multiplayer` was right, you are breaking this, also for the code bellow you should add a size check for the sound class blocks if possible

Reply:
https://github.com/pnill/cartographer/pull/407/commits/15ddad208effb18819e21dedf764dd9bbc93430a#diff-53f3f1a9f579c86f48f5e72e7355f36eddde2a46ee2fed9289c1410426dba388R56
Fixed